src/Common.Lib/systemstring.cpp

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "stdafx.h"
+#include "systemstring.h"
+#include <vector>
+#include <memory>
+
+#ifdef PLATFORM_UNIX
+#include <errno.h>
+#endif
+
+using namespace std;
+
+namespace CommonLib
+{
+#ifdef PLATFORM_UNIX
+
+    static const SIZE_T MAX_STRING_LEN = 10000;
+    struct Ciconv {
+        iconv_t m_iconv;
+        Ciconv(const iconv_t& iconv) : m_iconv(iconv) {}
+        ~Ciconv()
+        {
+            iconv_close(m_iconv);
+        }
+    };
+
+    HRESULT ErrnoToHResult(int& err) 
+    {
+        HRESULT hr;
+        switch (err)
+        {
+            case 0:
+                return S_OK;
+            case E2BIG:
+                return E_BOUNDS;
+            case EILSEQ:
+                return E_INVALIDARG;
+            case EINVAL:
+                return E_INVALIDARG;
+        }
+        return E_FAIL;
+    }
+
+    SystemString::SystemString(_In_ const tstring& original)
+    {
+        m_hresult = FromString(original.c_str(), *this);
+    }
+
+    SystemString::SystemString(_In_ const WCHAR* lpzwStr)
+    {
+        m_hresult = FromString(lpzwStr, *this);
+    }
+
+    HRESULT SystemString::ToString(_Inout_ tstring& str)
+    {
+        return Convert(c_str(), str);
+    }
+
+    HRESULT SystemString::ToString(_Inout_ string& str)
+    {
+        str = *this;
+        return S_OK;
+    }
+
+    HRESULT SystemString::FromString(_In_ const WCHAR* lpzwStr, _Inout_ SystemString& result)
+    {
+        if (lpzwStr == nullptr)
+        {
+            result = "";
+            result.m_hresult = E_INVALIDARG;
+            return result.m_hresult;
+        }
+
+        size_t i = 0;
+        const WCHAR* p = lpzwStr;
+        // scan to a resonable length for the end of the string.
+        for (i = 0; i < MAX_STRING_LEN; ++i, ++p)
+        {
+            if (*p == (WCHAR)0)
+            {
+                break;
+            }
+        }
+
+        if (i > MAX_STRING_LEN)
```

Comment:
>> [](http://example.com/codeflow?start=14&length=1)

`==` ? Otherwise, the for-loop condition will never make this true

Reply:
maybe `if (*p != (WCHAR)0)`

Reply:
Forgot to push this change. Checking for not null is probably a bad way to do this, because it could be null by coincidence.